### PR TITLE
#7571 Correctly process links with query parameters

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

**Review by rymsha — CHANGES_REQUESTED** [review]
this is not according to spec.

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/url/HtmlLinkProcessor.java [hunk 1]
```diff
@@ -278,4 +282,31 @@ private Map<String, String> extractUrlParams( final String urlQuery )
...
+            return url;
+        }
+        final String query = urlQuery.startsWith( "?" ) ? urlQuery.substring( 1 ) : urlQuery;
+
+        final StringBuilder urlSuffix = new StringBuilder();
+        final Map<String, String> queryParamsAsMap = extractUrlParams( query );
+
+        if ( queryParamsAsMap.containsKey( "query" ) )
+        {
+            urlSuffix.append( "?" ).append( decodeParameterValue( queryParamsAsMap.get( "query" ) ) );
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/url/HtmlLinkProcessor.java:298]: you should check that query does not contain illegal characters after `decodeParameterValue( queryParamsAsMap.get( "query" )`

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/url/HtmlLinkProcessor.java [hunk 2]
```diff
@@ -278,4 +282,31 @@ private Map<String, String> extractUrlParams( final String urlQuery )
...
+        }
+        if ( queryParamsAsMap.containsKey( "fragment" ) )
+        {
+            urlSuffix.append( "#" ).append( decodeParameterValue( queryParamsAsMap.get( "fragment" ) ) );
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/url/HtmlLinkProcessor.java:302]: you should check that query does not contain illegal characters after `decodeParameterValue( queryParamsAsMap.get( "fragment" )`

File: modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/url/HtmlLinkProcessor.java [hunk 3]
```diff
@@ -122,7 +125,8 @@ public String process( final String text, final String urlType, final PortalRequ
                             type( urlType ).
                             id( id ).
                             portalRequest( portalRequest );
-                        final String pageUrl = portalUrlService.pageUrl( pageUrlParams );
+                        final String pageUrl = addQueryParamsIfPresent( portalUrlService.pageUrl( pageUrlParams ), urlParamsString );
+
```
Comment (rymsha) [inline modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/url/HtmlLinkProcessor.java:129]: resulting pageUrl must be html (attribute) encoded

**Review by rymsha — CHANGES_REQUESTED** [review]

**Review by vbradnitski — APPROVED** [review]

**Pull request merged**